mountpoint-s3-crt/src/auth/imds_client.rs

```diff
@@ -0,0 +1,119 @@
+//! A client for retrieving ec2 instance metadata.
+
+use crate::auth::auth_library_init;
+use crate::common::allocator::Allocator;
+use crate::common::error::Error;
+use crate::io::channel_bootstrap::ClientBootstrap;
+use crate::io::retry_strategy::RetryStrategy;
+use crate::CrtError;
+use mountpoint_s3_crt_sys::{
+    aws_byte_buf, aws_imds_client, aws_imds_client_get_instance_type, aws_imds_client_new, aws_imds_client_options,
+    aws_imds_client_release,
+};
+use std::ptr::NonNull;
+
+/// A client for instance metadata query.
+#[derive(Debug)]
+pub struct ImdsClient {
+    /// A pointer to the underlying `aws_imds_client`
+    inner: NonNull<aws_imds_client>,
+}
+
+/// Configurations for creating a new [ImdsClient].
+#[derive(Debug, Default)]
+pub struct ImdsClientConfig {
+    /// The structure that can pass into the CRT's functions.
+    inner: aws_imds_client_options,
+
+    /// The [ClientBootstrap] to use to create connection to IMDS.
+    client_bootstrap: Option<ClientBootstrap>,
+
+    /// The [RetryStrategy] to use to reschedule failed requests to IMDS. This is reference counted,
+    /// so we only need to hold onto it until this [ImdsClientConfig] is consumed, at which point the
+    /// client will take ownership.
+    retry_strategy: Option<RetryStrategy>,
+}
+
+impl ImdsClient {
+    /// Create a new [ImdsClient]
+    pub fn new(allocator: &Allocator, config: ImdsClientConfig) -> Result<Self, Error> {
+        auth_library_init(allocator);
+
+        // SAFETY: `config` is moved into the [ImdsClient] on success, so `config.inner` (and the values
+        // inside) are guaranteed to live at least as long as this [ImdsClient] does. `allocator` is
+        // guaranteed to be a valid allocator because of the type-safe wrapper.
+        let inner = unsafe { aws_imds_client_new(allocator.inner.as_ptr(), &config.inner).ok_or_last_error()? };
+
+        Ok(Self { inner })
+    }
+
+    /// Make a get instance type request to ec2 hosts using [ImdsClient].
+    pub fn get_instance_type<F>(&self, callback: F) -> Result<i32, Error>
```

Comment:
Oh, I haven't aware of this piece of code exists.
```
/// Some CRT functions return an int that is either AWS_OP_SUCCESS or AWS_OP_ERR, and the caller           
/// should use last_error to find out what happened. This simplifies that pattern.                         
impl CrtError for i32 {                                                                                    
    type Return = ();                                                                                      
                                                                                                           
    unsafe fn ok_or_last_error(self) -> Result<Self::Return, Error> {                                      
        match self {                                                                                       
            AWS_OP_SUCCESS => Ok(()),                                                                      
            AWS_OP_ERR => Err(Error::last_error()),                                                        
            // This case shouldn't happen if used correctly since we should use this on functions          
            // that only return SUCCESS or ERR. But if it does happen, we can attempt to convert the       
            // error code directly, which may or may not work (but at least the Error won't be swallowed). 
            n => Err(common::error::Error::from(n)),                                                       
        }                                                                                                  
    }                                                                                                      
}                                                                                                          

```